nixos/modules/services/torrent/transmission.nix

```diff
@@ -351,14 +363,6 @@ in
             cfg.settings.script-torrent-done-filename ++
           optional (cfg.settings.watch-dir-enabled && !cfg.settings.trash-original-torrent-files)
             cfg.settings.watch-dir;
-        StateDirectory = [
-          "transmission"
-          "transmission/${settingsDir}"
-          "transmission/${incompleteDir}"
-          "transmission/${downloadsDir}"
-          "transmission/${watchDir}"
-        ];
-        StateDirectoryMode = mkDefault 750;
```

Comment:
Please do not remove `StateDirectoryMode` nor `StateDirectory`, otherwise they will not be bound in the `RootDirectory`.

Reply:
I don't exactly understand the bind mount stuff from the systemd docs. Isn't it enough to only bind the top-level directory (`/var/lib/transmission`)?

Reply:
yes, but listing subdirs [enforces `StateDirectoryMode` on them](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html):

> Except in case of ConfigurationDirectory=, the innermost specified directories will be owned by the user and group specified in User= and Group=. If the specified directories already exist and their owning user or group do not match the configured ones, all files and directories below the specified directories as well as the directories themselves will have their file ownership recursively changed to match what is configured. As an optimization, if the specified directories are already owned by the right user and group, files and directories below of them are left as-is, even if they do not match what is requested. The innermost specified directories will have their access mode adjusted to the what is specified in RuntimeDirectoryMode=, StateDirectoryMode=, CacheDirectoryMode=, LogsDirectoryMode= and ConfigurationDirectoryMode=.

Note that because `"transmission/${downloadsDir}"` is necessarily `/var/lib/transmission/Downloads`, `StateDirectoryMode` does not adjust modes of `cfg.settings.download-dir` if it's set to something else than the default `"${cfg.home}/${downloadsDir}"` (`cfg.home` being another quirk that lingers).

Reply:
I see. Well IMO this entire module deserves are a rewrite, but that's a lot of effort, so the latest commit should be fine.

Reply:
Yes, if it were not for backward compatibility we should drop `cfg.home` to always use `/var/lib/transmission`, and drop `downloadDirPermissions` to let `systemd` create directories with "750" (or "770" if members of the `transmission` group should be allowed to delete dangling torrent files that transmission did not delete).

Except for that, the rest of the module is not that bad (obviously I'm biased),
amongst the things that come to mind that can be improved security-wise:
1. With recent `systemd` giving access to `LoadCredential=`/`LoadCredentialEncrypted=` in `ExecStartPre=`, those should be used instead of directly using `${cfg.credentialsFile}`, as in https://github.com/NixOS/nixpkgs/blob/072f2eef83e8bd45be52ea6fdbb8ea488867a945/nixos/modules/services/misc/radicle.nix#L272-L282
2. Instead of using `BindReadOnlyPaths = [ builtins.storeDir ]`, `systemd.services.transmission.confinement` should be used to only mount transmission's closure dependencies, as in https://github.com/NixOS/nixpkgs/blob/072f2eef83e8bd45be52ea6fdbb8ea488867a945/nixos/modules/services/misc/radicle.nix#L102-L112